FFmpegInterop/FFmpegInteropMSS.h

```diff
@@ -17,148 +17,56 @@
-				return mediaDuration;
-			};
-		};
-		property String^ VideoCodecName
-		{
-			String^ get()
-			{
-				return videoCodecName;
-			};
-		};
-		property String^ AudioCodecName
-		{
-			String^ get()
-			{
-				return audioCodecName;
-			};
-		};
-		property String^ SubtitleCodecName
-		{
-			String^ get()
-			{
-				return subtitleCodecName;
-			};
-		};
-
-		void ReleaseFileStream();
+	private:
+		FFmpegInteropMSS(_In_ const Windows::Media::Core::MediaStreamSource& mss);
 
-	internal:
-		int ReadPacket();
+		void OpenFile(_In_ const Windows::Storage::Streams::IRandomAccessStream& fileStream, _In_opt_ const FFmpegInterop::FFmpegInteropMSSConfig& config);
+		void OpenFile(_In_z_ const char* uri, _In_opt_ const FFmpegInterop::FFmpegInteropMSSConfig& config);
 
-	private:
-		FFmpegInteropMSS();
+		void InitFFmpegContext(_In_opt_ const FFmpegInterop::FFmpegInteropMSSConfig& config);
 
-		HRESULT CreateMediaStreamSource(IRandomAccessStream^ stream, bool forceAudioDecode, bool forceVideoDecode, PropertySet^ ffmpegOptions, MediaStreamSource^ mss);
-		HRESULT CreateMediaStreamSource(String^ uri, bool forceAudioDecode, bool forceVideoDecode, PropertySet^ ffmpegOptions);
-		HRESULT InitFFmpegContext(bool forceAudioDecode, bool forceVideoDecode);
-		HRESULT CreateAudioStreamDescriptor(bool forceAudioDecode);
-		HRESULT CreateAudioStreamDescriptorFromParameters(const AVCodecParameters* avCodecParams);
-		HRESULT CreateVideoStreamDescriptor(bool forceVideoDecode);
-		HRESULT CreateSubtitleStreamDescriptor(const AVStream* avStream);
-		HRESULT ConvertCodecName(const char* codecName, String^ *outputCodecName);
-		HRESULT ParseOptions(PropertySet^ ffmpegOptions);
-		void OnStarting(MediaStreamSource ^sender, MediaStreamSourceStartingEventArgs ^args);
-		void OnSampleRequested(MediaStreamSource ^sender, MediaStreamSourceSampleRequestedEventArgs ^args);
-		void OnSwitchStreamsRequested(MediaStreamSource ^sender, MediaStreamSourceSwitchStreamsRequestedEventArgs ^args);
+		void OnStarting(_In_ const Windows::Media::Core::MediaStreamSource& sender, _In_ const Windows::Media::Core::MediaStreamSourceStartingEventArgs& args);
+		void OnSampleRequested(_In_ const Windows::Media::Core::MediaStreamSource& sender, _In_ const Windows::Media::Core::MediaStreamSourceSampleRequestedEventArgs& args);
+		void OnSwitchStreamsRequested(_In_ const Windows::Media::Core::MediaStreamSource& sender, _In_ const Windows::Media::Core::MediaStreamSourceSwitchStreamsRequestedEventArgs& args);
+		void OnClosed(_In_ const Windows::Media::Core::MediaStreamSource& sender, _In_ const Windows::Media::Core::MediaStreamSourceClosedEventArgs& args);
 
-		MediaStreamSource^ mss;
-		EventRegistrationToken startingRequestedToken;
-		EventRegistrationToken sampleRequestedToken;
-		EventRegistrationToken switchStreamsRequestedToken;
+		std::mutex m_lock;
+		Windows::Media::Core::MediaStreamSource m_mss; // We hold a circular reference to the provided MSS which we break when the Closed event is fired

```

Comment:
On closer inspection, it looks like this is specific to the new implementation as we now explicitly hold a strong reference to FFmpegInteropMSS in the delegates:

`// Register event handlers. The delegates hold strong references to tie the lifetime of this object to the MSS.`
`m_startingEventToken = m_mss.Starting({ get_strong(), &FFmpegInteropMSS::OnStarting });`

Reply:
Lukas is correct that the events were weak references before so there was no circular reference. I made them strong references using get_strong() so that the FFmpegInteropMSS is guaranteed to be kept alive as long as it has events registered with the MSS.

There's a good C++/WinRT article covering this topic:
https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/weak-references

Reply:
Actually, after I wrote my comment, I thought that it would be cool if this was a strong reference, for the reasons @brbeec mentioned. I will try to get this change also into our fork, just need to find a way to do it in C++/CX.